main.cpp

```diff
@@ -263,75 +167,120 @@ struct HeapA
 };
 
 #include <iostream>
+#include <cmath>
 
 struct FloatType;
 struct DoubleType;
 struct IntType;
 
+//UDTs
+
 //FloatType
 struct FloatType
 {
-    FloatType(float v) : value(new float(v)) {}
-
-    ~FloatType()
-    {
-        delete value;
-    }
+    explicit FloatType(float v);
+    ~FloatType();
 
     FloatType& add(float x);
     FloatType& subtract(float x);
     FloatType& multiply(float x);
     FloatType& divide(float x);
+    FloatType& pow(float y);
+    FloatType& pow(const FloatType& y);
+    FloatType& pow(const DoubleType& y);
+    FloatType& pow(const IntType& y);
 
     operator float() const { return *value; }
 
 private:
     float* value = nullptr;
+    FloatType& powInternal(float arg);
 };
 
 //DoubleType
 struct DoubleType
 {
-    DoubleType(double v) : value(new double(v)) {}
-
-    ~DoubleType()
-    {
-        delete value;
-    }
+    explicit DoubleType(double v);
+    ~DoubleType();
 
     DoubleType& add(double x);
     DoubleType& subtract(double x);
     DoubleType& multiply(double x);
     DoubleType& divide(double x);
+    DoubleType& pow(double y);
+    DoubleType& pow(const FloatType& y);
+    DoubleType& pow(const DoubleType& y);
+    DoubleType& pow(const IntType& y);
 
     operator double() const { return *value; }
 
 private:
     double* value = nullptr;
+    DoubleType& powInternal(double arg);
 };
 
 //IntType
 struct IntType
 {
-    IntType(int v) : value(new int(v)) {}
-
-    ~IntType()
-    {
-        delete value;
-    }
+    explicit IntType(int v);
+    ~IntType();
 
     IntType& add(int x);
     IntType& subtract(int x);
     IntType& multiply(int x);
     IntType& divide(int x);
+    IntType& pow(int y);
+    IntType& pow(const FloatType& y);
+    IntType& pow(const DoubleType& y);
+    IntType& pow(const IntType& y);
 
     operator int() const { return *value; }
 
 private:
     int* value = nullptr;
+    IntType& powInternal(int arg);
 };
 
-//functions
+//UDT functions
+
+//float
+FloatType::FloatType(float v) : value(new float(v)) {}
+
+FloatType::~FloatType() 
+{
+    delete value;
+}
+
+FloatType& FloatType::pow(float y)
+{
+    powInternal(y);
+    return *this;
+}
+
+FloatType& FloatType::pow(const FloatType& y)
+{
+    powInternal(static_cast<float>(y));
+    return *this;
+}
+
+FloatType& FloatType::pow(const DoubleType& y)
+{
+    powInternal(static_cast<float>(y));
+    return *this;
+}
+
+FloatType& FloatType::pow(const IntType& y)
+{
+    powInternal(static_cast<float>(y));
+    return *this;
+}
```

Comment:
`powInternal` returns `*this` already. 
you don't need a separate return statement.
```
return powInternal(...);
```
- [ ] revise